wrappers/kmip/kmip.go

```diff
@@ -0,0 +1,537 @@
+	EnvKmipCaCert       = "KMIP_CA_CERT"
+	EnvKmipClientCert   = "KMIP_CLIENT_CERT"
+	EnvKmipClientKey    = "KMIP_CLIENT_KEY"
+	EnvKmipServerName   = "KMIP_SERVER_NAME"
+	EnvKmipTimeout      = "KMIP_TIMEOUT"
+	EnvKmipEncryptAlg   = "KMIP_ENCRYPT_ALG"
+	EnvKmipTls12Ciphers = "KMIP_TLS12_CIPHERS"
+)
+
+var (
+	// Preset kmip cryptographic parameters that can be selected from config
+	cryptoParamsPreset = map[string]kmip.CryptographicParameters{
+		"AES_GCM": {
+			CryptographicAlgorithm: kmip.CryptographicAlgorithmAES,
+			BlockCipherMode:        kmip.BlockCipherModeGCM,
+			TagLength:              16,
+			IVLength:               12,
+		},
+		"RSA_OAEP_SHA256": {
+			CryptographicAlgorithm:        kmip.CryptographicAlgorithmRSA,
+			PaddingMethod:                 kmip.PaddingMethodOAEP,
+			HashingAlgorithm:              kmip.HashingAlgorithmSHA_256,
+			MaskGenerator:                 kmip.MaskGeneratorMGF1,
+			MaskGeneratorHashingAlgorithm: kmip.HashingAlgorithmSHA_256,
+		},
+		"RSA_OAEP_SHA384": {
+			CryptographicAlgorithm:        kmip.CryptographicAlgorithmRSA,
+			PaddingMethod:                 kmip.PaddingMethodOAEP,
+			HashingAlgorithm:              kmip.HashingAlgorithmSHA_384,
+			MaskGenerator:                 kmip.MaskGeneratorMGF1,
+			MaskGeneratorHashingAlgorithm: kmip.HashingAlgorithmSHA_384,
+		},
+		"RSA_OAEP_SHA512": {
+			CryptographicAlgorithm:        kmip.CryptographicAlgorithmRSA,
+			PaddingMethod:                 kmip.PaddingMethodOAEP,
+			HashingAlgorithm:              kmip.HashingAlgorithmSHA_512,
+			MaskGenerator:                 kmip.MaskGeneratorMGF1,
+			MaskGeneratorHashingAlgorithm: kmip.HashingAlgorithmSHA_512,
+		},
+	}
+
+	// Map between crypto param name, and mechanism ID. The ID is saved with the wrapped key
+	// so we can use the same for decryption. This allows changing the cryptographic parameter for the next encryption.
+	mechanisms = []string{"AES_GCM", "RSA_OAEP_SHA256", "RSA_OAEP_SHA384", "RSA_OAEP_SHA512"}
+)
+
+// Wrapper is a Wrapper that uses KMIP
+type Wrapper struct {
+	// keyId is the private key or symmetric key ID.
+	keyId string
+	// pubKeyId is used for asymmetric encryption.
+	// It's automatically discovered and set during initialization.
+	// It's left empty when using a symmetric key / algorithm.
+	pubKeyId     string
+	endpoint     string
+	timeout      uint64
+	opts         []kmipclient.Option
+	cryptoParams kmip.CryptographicParameters
+	// This is a id saved into the encrypted payload next to the decryption key ID so we can reuse the same
+	// algorithm for decryption that was used for encryption. This allows for changing the algorithm while keeping
+	// the capability to decrypt previously encrypted values.
+	mechanismID uint64
+	// cachedClient is a closed kmip connection that can be reopened.
+	// It will reuse initial connection parameters and protocol version
+	// negociated the first time. This avoids useless verion negociations on subsequent connections.
```

Comment:
I think this makes sense. In PKCS#11 world, we leave open initialization (which maybe leaves a long-lived connection around), but open a new session per request.